public/app/core/logsModel.ts

```diff
@@ -359,39 +337,27 @@ export function logSeriesToLogsModel(logSeries: DataFrame[], queries: DataQuery[
   const allLabels: Labels[][] = [];
 
   // Find the fields we care about and collect all labels
-  let allSeries: LogFields[] = [];
+  let allSeries: LogInfo[] = [];
 
   // We are sometimes passing data frames with no fields because we want to calculate correct meta stats.
   // Therefore we need to filter out series with no fields. These series are used only for meta stats calculation.
   const seriesWithFields = logSeries.filter((series) => series.fields.length);
 
   if (seriesWithFields.length) {
     seriesWithFields.forEach((series) => {
-      const fieldCache = new FieldCache(series);
-      const stringField = fieldCache.getFirstFieldOfType(FieldType.string);
-      const timeField = fieldCache.getFirstFieldOfType(FieldType.time);
-      // NOTE: this is experimental, please do not use in your code.
-      // we will get this custom-frame-type into the "real" frame-type list soon,
-      // but the name might change, so please do not use it until then.
-      const labelsField =
-        series.meta?.custom?.frameType === 'LabeledTimeValues' ? fieldCache.getFieldByName('labels') : undefined;
-
-      if (stringField !== undefined && timeField !== undefined) {
+      const logsFrame = parseLogsFrame(series);
+      if (logsFrame != null) {
+        // for now we ignore the nested-ness of attributes, and just stringify-them
+        const frameLabels = logsFrame.attributes?.map(attributesToLabels);
         const info = {
-          series,
-          timeField,
-          labelsField,
-          timeNanosecondField: fieldCache.getFieldByName('tsNs'),
-          stringField,
-          logLevelField: fieldCache.getFieldByName('level'),
-          idField: getIdField(fieldCache),
+          rawFrame: series,
+          logsFrame: logsFrame,
+          frameLabels,
         };
 
         allSeries.push(info);
-
-        const labels = getAllLabels(info);
-        if (labels.length > 0) {
-          allLabels.push(labels);
+        if (frameLabels != null && frameLabels.length > 0) {
```

Comment:
I think based on https://github.com/grafana/grafana/pull/70318/files#diff-07da85e1ede94286e302d4831fd8768e312dd25291a32fb9bfb1f02115331121R50 `frameLabels` can also be `undefined`? I think it's better to check agains `!frameLabels` anyways.

Reply:
this will be a long discussion :) 

basically, `x != null` is the same as `(x !== null) && (x !== undefined)`. i nearly always write `!= null`, regardless of whether i am comparing for `null` or `undefined`. it's a trick, and maybe tricks are bad.

to clarify, is your recommendation to go with
```js
if (frameLabels && frameLabels.length > 0)
```
?